runtime/metricsview/executor.go

```diff
@@ -65,10 +67,46 @@ func (e *Executor) Close() {
 	e.olapRelease()
 }
 
-// Cacheable returns whether the result of running the given query is cacheable.
-func (e *Executor) Cacheable(qry *Query) bool {
-	// TODO: Get from OLAP instead of hardcoding
-	return e.olap.Dialect() == drivers.DialectDuckDB
+// CacheKey returns a cache key based on the executor's metrics view's cache key configuration.
+// If ok is false, caching is disabled for the metrics view.
+func (e *Executor) CacheKey(ctx context.Context) ([]byte, bool, error) {
+	spec := e.metricsView
+	// Cache is disabled for metrics views based on external table
+	if (spec.CacheEnabled != nil && !*spec.CacheEnabled) || (spec.CacheEnabled == nil && e.streaming) {
+		return nil, false, nil
+	}
+
+	if spec.CacheKeySql == "" {
+		if !e.streaming {
+			// for metrics views on rill managed tables no cache key specific to the metrics view is required
+			return []byte(""), true, nil
+		}
+		// watermark is the default cache key for streaming metrics views
+		watermark, err := e.loadWatermark(ctx, nil)
+		if err != nil {
+			return nil, false, err
+		}
+		return []byte(watermark.Format(time.RFC3339)), true, nil
+	}
+
+	res, err := e.olap.Execute(ctx, &drivers.Statement{
+		Query:    spec.CacheKeySql,
+		Priority: e.priority,
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	defer res.Close()
+	var key string
+	for res.Next() {
+		if err := res.Scan(&key); err != nil {
+			return nil, false, err
+		}
+	}
+	if res.Err() != nil {
+		return nil, false, err
+	}
+	return []byte(key), true, nil
```

Comment:
Not sure if we want to protect agains this, but often the key will be a timestamp and people may forget to cast it to a string. 

An easy way to protect here is to use `var key any` and then run it through `jsonval.ToValue` and then `json.Marshal` – that will almost always give you a safely serialized string representation.

Reply:
Weirdly I did not find this issue in my testing and on digging further I realise that the `sql` package already does conversion when source is `time.Time` and dest is `*string` : https://github.com/golang/go/blob/669d87a935536eb14cb2db311a83345359189924/src/database/sql/convert.go#L286
It does this conversions for most common types.

But I think it is good to run it through `jsonval.ToValue` since it handles more types.